src/approximation_methods.jl

```diff
@@ -0,0 +1,93 @@
+@doc raw"""
+    AbstractApproximationMethod
+
+Abstract type for defining estimation methods on manifolds.
+"""
+abstract type AbstractApproximationMethod end
+
+@doc raw"""
+    GradientDescentEstimation <: AbstractApproximationMethod
+
+Method for estimation using [📖 gradient descent](https://en.wikipedia.org/wiki/Gradient_descent).
+"""
+struct GradientDescentEstimation <: AbstractApproximationMethod end
+
+@doc raw"""
+    CyclicProximalPointEstimation <: AbstractApproximationMethod
+
+Method for estimation using the cyclic proximal point technique, which is based on [📖 proximal maps](https://en.wikipedia.org/wiki/Proximal_operator).
+"""
+struct CyclicProximalPointEstimation <: AbstractApproximationMethod end
+
+@doc raw"""
+    EfficientEstimator <: AbstractApproximationMethod
+
+Method for estimation in the best possible sense, see [📖 Efficiency (Statictsics)](https://en.wikipedia.org/wiki/Efficiency_(statistics)) for mroe details.
+This can for example be used when computing the usual mean on an Euclidean space, which is the best estimator.
+"""
+struct EfficientEstimator <: AbstractApproximationMethod end
+
+
+@doc raw"""
+    ExtrinsicEstimation{T} <: AbstractApproximationMethod
+
+Method for estimation in the ambient space with a method of type `T` and projecting the result back
+to the manifold.
+"""
+struct ExtrinsicEstimation{T} <: AbstractApproximationMethod
+    extrinsic_estimation::T
+end
+
+@doc raw"""
+    WeiszfeldEstimation <: AbstractApproximationMethod
+
+Method for estimation using the Weiszfeld algorithm, compare for example the computation of the
+[📖 Geometric median](https://en.wikipedia.org/wiki/Geometric_median).
+"""
+struct WeiszfeldEstimation <: AbstractApproximationMethod end
+
+@doc raw"""
+    GeodesicInterpolation <: AbstractApproximationMethod
+
+Method for estimation based on geodesic interpolation.
+"""
+struct GeodesicInterpolation <: AbstractApproximationMethod end
+
+@doc raw"""
+    GeodesicInterpolationWithinRadius{T} <: AbstractApproximationMethod
+
+Method for estimation based on geodesic interpolation that is restricted to some `radius`
+
+# Constructor
+
+    GeodesicInterpolationWithinRadius(radius)
+"""
+struct GeodesicInterpolationWithinRadius{T} <: AbstractApproximationMethod
+    radius::T
+    function GeodesicInterpolationWithinRadius(radius::T) where {T}
+        radius > 0 && return new{T}(radius)
+        return throw(
+            DomainError("The radius must be strictly postive, received $(radius)."),
+        )
+    end
+end
+
+@doc raw"""
+    default_estimation_method(M::AbstractManifold)
+    default_estimation_method(M::AbtractManifold, f, T)
+
+Specify a default estimation method for an [`AbstractManifold`](@ref) and (optional)
+for a specific function `f` and a type `T` to distinguish different (point or vector)
+representations on M.
+
+By default, all functions `f` call the signature for just a manifold.
```

Comment:
If you have a manifold default, like: We do everything here with gradient stuff.

For example one could implement the mean for general manifold with different methods in general and then say something like: Sphere? Gradient! So its gradient. Hyperbollic? CPPA!

Reply:
Hm. I would prefer to leave that for a future PR/separate discussion.

Reply:
Hm, we do that in a few other places as well like the objectivity radius

Reply:
So for now an easy (lazy) reason is in testing – I just set the default approximation for every function on the `DefaultManifold` to gradient descent and can check all function fallbacks :)

Reply:
Well, approximation methods for retract are supposed to be `<:AbstractRetractionMethod`, for inverse retract `<:AbstractInverseRetractionMethod`, so this alone tells me that we can't have the same method for all functions.

Reply:
but they also never fall back to the generic default, since `default_approximation_method(M,::retract) is always implemented.

Reply:
Then for which functions we need that default? You seem to be only testing for `manifold_dimension` but we don't have any manifold for which we would have trouble with exact calculation of its dimension.

Reply:
The is just laziness and I took the first function I could come up with.

The realistic scenario (but we do not have those functions here) is that often both median and mean have the same default method to be computed with on a certain manifold – probably other statistic functions then as well.

Reply:
And I am also not saying this it _the_ feature that is completely necessary – but it is a nice generic fallback we could provide for whoever needs it.

Reply:
Isn't mean usually computed using GD, while median using some proximal algorithm like CPPA? Anyway, if someone wants to do that for their manifold, they can implement `default_approximation_method(M::MyManifold, f) = default_approximation_method(M)`. IMO doing it in general would lead to confusing messages about method X not working for approximation of Y.

Reply:
Ok. Your solution sounds fair enough as well.

Reply:
Ah, there is one problem with your idea; ambiguities. Since we have `default_approximation_method(M::AbstractManifolds, ::typeof(retract)) = ...`
actually defining `default_approximation_method(::MyFavouriteManifold, f) = ...` is ambiguous and not (easily) possible.

Reply:
I still removed the global default to not have an infinite discussion – also this way it is easy to add this later if we feel that is still more useful (which is easier than removing it later ;))

Reply:
Right, but these ambiguities can be fairly easily resolved.

Reply:
Well, it basically means one can not provide a default for all f that way.

Reply:
Unless they also define `default_approximation_method(::MyFavouriteManifold, f::typeof(retract)) = ...` and so on, which is not much work.

Anyway, thanks for removing this default from this PR; we can always introduce it later when tradeoffs are carefully considered.